File: 540 Team Queue/solution.java
import java.io.*;
import java.util.*;

public class Main {
	public static void main(String[] args) throws IOException {
        int[] teamNum = new int[1000000];
        StreamTokenizer st = new StreamTokenizer(System.in);
        int cases = 0, t;
        StringBuilder output = new StringBuilder();
        while(st.nextToken() == StreamTokenizer.TT_NUMBER && (t = (int)st.nval) != 0) {
    	    output.append("Scenario #").append(++cases).append('\n');
    		for(int i = 0; i < t; ++i) {
    			st.nextToken();
    			int elementCount = (int)st.nval;
    			while(elementCount-- > 0) {
    			    st.nextToken();
    				teamNum[(int)st.nval] = i;
    			}
    		}
            ArrayList<Integer> queue = new ArrayList<>();
            int next = 0;
		    String command;
    	    while(st.nextToken() == StreamTokenizer.TT_WORD && (command = st.sval).charAt(0) != 'S') {
    	        if(command.charAt(0) == 'E') {
    			    st.nextToken();
    			    queue.add((int)st.nval);
    	        } else if(command.charAt(0) == 'D') {
    			    int dequeue = queue.get(next);
    			    output.append(dequeue).append('\n');
    			    queue.remove(next);
    			    int currTeam = teamNum[dequeue];
    			    while(next < queue.size() && teamNum[queue.get(next)] != currTeam)
    			        ++next;
    			    if(next == queue.size())
    			        next = 0;
    			}
    	    }
    	    output.append('\n');
        }
	    System.out.print(output);
	}
}

File: 540 Team Queue/solution.c
#include <stdio.h>
#include <stdlib.h>

int teamNum[1000000];

typedef struct Queue {
    int data;
    struct Queue *prev, *next;
} Queue;

int main() {
	int cases = 0, t;
	while(scanf("%d", &t) && t != 0) {
	    printf("Scenario #%d\n", ++cases);
	    int elementCount;
	    for(int i = 0; i < t; ++i) {
	        scanf("%d", &elementCount);
	        while(elementCount--) {
	            int element;
	            scanf("%d", &element);
	            teamNum[element] = i;
	        }
	    }
	    Queue *head = NULL, *tail = NULL, *nextDequeue = NULL;
	    char command[8];
	    while(scanf("%s", command) && command[0] != 'S') {
	        if(command[0] == 'E') {
	            Queue *newNode = (Queue*)malloc(sizeof(Queue));
	            scanf("%d", &newNode->data);
	            newNode->prev = tail;
	            newNode->next = NULL;
	            if(!head)
	                head = nextDequeue = newNode;
	            else
	                tail->next = newNode;
	            tail = newNode;
	        } else if(command[0] == 'D') {
	            printf("%d\n", nextDequeue->data);
	            int currTeam = teamNum[nextDequeue->data];
	            Queue *temp;
	            if(nextDequeue == head) {
	                temp = head;
	                head = head->next;
	            } else if(nextDequeue == tail) {
	                temp = tail;
	                tail = tail->prev;
	                tail->next = NULL;
	            } else {
	                temp = nextDequeue;
	                nextDequeue->prev->next = nextDequeue->next;
	                nextDequeue->next->prev = nextDequeue->prev;
	            }
	            free(temp);
	            do {
	                nextDequeue = nextDequeue->next;
	            } while(nextDequeue && teamNum[nextDequeue->data] != currTeam);
	            if(!nextDequeue)
	                nextDequeue = head;
	        }
	    }
	    putchar('\n');
	    while(head) {
	        Queue *temp = head;
	        head = head->next;
	        free(temp);
	    }
	}
	return 0;
}
